text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+A `--no-resolve-sysroot` flag will be added for this purpose.
+Compilers that don't have sysroot binaries should accept this flag and ignore it.
+Once [Rust PR #35021](https://github.com/rust-lang/rust/pull/35021/files) lands in some form, `compiler-rt` will be a Cargoized dependency so rustc won't use the sysroot for linking either.
+
+Additionally, compilers besides rustc may have version numbers distinct from the version of Rust they implement.
+For this purpose, the verbose version output (`$COMPILER -vV` should contain an additional line:
+```
+language-version: $version
+```
+For now, `$version` should be a version, not a version requirement, and the patch number must be zero as patch numbers don't make sense for interfaces.
+[This may be relaxed in the future for compilers which implement multiple versions.]
+
+## Compiler source packaging
+
+While the standard library *interface* is defined with each rustc version, the implementation of many crates, by virtue of using unstable features, is compiler-specific.
+This makes the standard library unfit for crates.io.
+(Additionally, the issue of dealing with nightly also makes crates.io hard to use, but that is a less clear-cut obstacle.)
+
+Cargo will soon gain the ability to create custom registries, as described in
+https://github.com/rust-lang/cargo/pull/2857 .
+Compilers should package the source of their implementation of the standard library as a registry, which can be distributed with the compiler.
+In practice, rustc will optionally contain the source in its sysroot.
+Rustup may be able to put it there if the default download does not contain it already.
+This registry will be called the "compiler source registry".
+
+## Cargo implementation
+
+Whenever Cargo encounters a `Cargo.toml`, the first thing it always does is inject any applicable implicit deps.
+The idea is that by doing this so early on, most of Cargo can stay the same in only knowing or caring about explicit deps, simplifying both this RFC and its implementation.
+This process is completely defined by the rules described in the first subsection, so there is really nothing to elaborate upon here.
+
+Cargo has a "source" abstraction for providers of packages.
+Examples of this are file-system paths, git repositories, and the upcoming registries.
+For the time being, `stdlib = true` will act like another source, instructing the compiler to look at the compiler source registry, and if that is missing, then the sysroot binaries.
+The remainder of this section will clarify this.
+
+The compiler source registry is expected to be located in `${$CARGO_RUSTC --print sysroot}/src`.
+The exact format this takes will be determined when custom registries are implemented.
+Based on that, it may be necessary to transform the rust repo before dumping it there.
+
+The "sysroot binary mock registry" is generated by examining the contents of the sysroot.
```

Comment:
Let's be more precise than 'contents of the sysroot'. It will look in `$sysroot/lib/rustlib/$target-triple/lib`.